### PR TITLE
Fix/#3184/fix server crashs on database change

### LINKED ISSUE #3184: Changing default database name crashes server
**Informations**
- **Node.js version**:  10.15.0
- **NPM version**:  6.4.1
- **Strapi version**:  latest cli install today 
- **Database**: PostgresSQL 11 local brew install
- **Operating system**:  MacOS latest
- **(Optional) Link to your Project**: 

**What is the current behavior?**  Server crashing when changing database name from default to anything else.  Unrecoverable, crashing restart also needs reinstall

**Steps to reproduce the problem**  

1. New manual install using local Postgresql
2.  Login new credentials
3.  Go to Configuration -> databases and change name from default to anything else
4.  On SAVE server crashes

```
[2019-04-24T21:27:18.828Z] debug PUT /settings-manager/configurations/databases/default/development (23 ms)
error TypeError: strapi.models.core_store.forge is not a function

error TypeError: strapi.models.core_store.forge is not a function
    at Object.get (/usr/local/lib/node_modules/strapi/lib/core/store.js:55:47)
    at module.exports (/Users/Numchucks/development/client/forms/cms/plugins/content-manager/config/functions/bootstrap.js:312:42)
    at Promise (/usr/local/lib/node_modules/strapi/lib/Strapi.js:299:13)
    at new Promise (<anonymous>)
    at execBootstrap (/usr/local/lib/node_modules/strapi/lib/Strapi.js:287:11)
    at Promise.all.Object.values.map.x (/usr/local/lib/node_modules/strapi/lib/Strapi.js:326:44)
    at Array.map (<anonymous>)
    at Strapi.bootstrap (/usr/local/lib/node_modules/strapi/lib/Strapi.js:326:35)
    at Strapi.start (/usr/local/lib/node_modules/strapi/lib/Strapi.js:109:18)
(node:95769) UnhandledPromiseRejectionWarning: TypeError: strapi.models.core_store.forge is not a function
    at Object.get (/usr/local/lib/node_modules/strapi/lib/core/store.js:55:47)
    at module.exports (/Users/Numchucks/development/client/forms/cms/plugins/settings-manager/config/functions/bootstrap.js:17:26)
    at Promise (/usr/local/lib/node_modules/strapi/lib/Strapi.js:299:13)
    at new Promise (<anonymous>)
    at execBootstrap (/usr/local/lib/node_modules/strapi/lib/Strapi.js:287:11)
    at Promise.all.Object.values.map.x (/usr/local/lib/node_modules/strapi/lib/Strapi.js:326:44)
    at Array.map (<anonymous>)
    at Strapi.bootstrap (/usr/local/lib/node_modules/strapi/lib/Strapi.js:326:35)
    at Strapi.start (/usr/local/lib/node_modules/strapi/lib/Strapi.js:109:18)
(node:95769) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 4)
(node:95769) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:95769) UnhandledPromiseRejectionWarning: TypeError: strapi.models.core_store.forge is not a function
    at Object.get (/usr/local/lib/node_modules/strapi/lib/core/store.js:55:47)
    at module.exports (/Users/Numchucks/development/client/forms/cms/plugins/users-permissions/config/functions/bootstrap.js:96:45)
    at Promise (/usr/local/lib/node_modules/strapi/lib/Strapi.js:299:13)
    at new Promise (<anonymous>)
    at execBootstrap (/usr/local/lib/node_modules/strapi/lib/Strapi.js:287:11)
    at Promise.all.Object.values.map.x (/usr/local/lib/node_modules/strapi/lib/Strapi.js:326:44)
    at Array.map (<anonymous>)
    at Strapi.bootstrap (/usr/local/lib/node_modules/strapi/lib/Strapi.js:326:35)
    at Strapi.start (/usr/local/lib/node_modules/strapi/lib/Strapi.js:109:18)
(node:95769) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 5)
(node:95769) UnhandledPromiseRejectionWarning: ReferenceError: config is not defined
    at fs.readdir (/Users/Numchucks/development/client/forms/cms/plugins/email/config/functions/bootstrap.js:68:40)
(node:95769) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 7)
(node:95769) UnhandledPromiseRejectionWarning: ReferenceError: config is not defined
    at fs.readdir (/Users/Numchucks/development/client/forms/cms/plugins/upload/config/functions/bootstrap.js:70:40)
(node:95769) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 9)
```

**What is the expected behavior?**

Change default database name

**Suggested solutions**

### PR DESCRIPTION
**Pull request by Convly**
#### Description of what you did:
fix  #3184 

It is now possible to remove the `connection` attribute from the model/component definition (`(model|component).settings.json`).
- If not set/present, then the default connection will be used (the one defined in `config/env/database.json` as `defaultConnection`)
- If set, the value must match one of the keys of the `connections` object in `config/env/database.json`

The connection attribute has been also removed from the models shipped with the strapi plugins (_admin, upload, permissions, etc..._).

#### Limitations
Relations between models & components cannot be set across models which have different connections

### CODE REVIEW

File: packages/strapi-generate-model/lib/before.js [hunk 1]
```diff
@@ -122,20 +122,6 @@ module.exports = (scope, cb) => {
     })
     .join(',\n');
 
-  // Get default connection
```
Comment (alexandrebodin) [inline packages/strapi-generate-model/lib/before.js:125]: You should just remove the JSON.parse part but allow a user to set a connection if wanted

File: packages/strapi-generate-model/templates/bookshelf/model.settings.template [hunk 2]
```diff
@@ -1,6 +1,5 @@
 {
   "kind": "collectionType",
-  "connection": "<%= connection %>",
```
Comment (alexandrebodin) [inline packages/strapi-generate-model/templates/bookshelf/model.settings.template:3]: Might need a refactor so it is set only if the connection arg is passed in the CLI

File: packages/strapi-generate-model/templates/mongoose/model.settings.template [hunk 3]
```diff
@@ -1,6 +1,5 @@
 {
   "kind": "collectionType",
-  "connection": "<%= connection %>",
```
Comment (alexandrebodin) [inline packages/strapi-generate-model/templates/mongoose/model.settings.template:3]: same here

File: packages/strapi/lib/Strapi.js [hunk 4]
```diff
@@ -318,9 +318,15 @@ class Strapi {
       configuration: this.config.get('currentEnvironment.server.webhooks', {}),
     });
 
+    const ensureDefaultConnection = model =>
+      _.merge(
+        { connection: this.config.get('currentEnvironment.database.defaultConnection') },
```
Comment (alexandrebodin) [inline packages/strapi/lib/Strapi.js:323]: can you instead just set the field in the model ?

File: packages/strapi/lib/core/bootstrap.js [hunk 5]
```diff
@@ -74,6 +74,19 @@ module.exports = function(strapi) {
     return acc;
   }, {});
 
+  // Set components
+  strapi.components = Object.keys(strapi.components).reduce((acc, componentName) => {
```
Comment (alexandrebodin) [inline packages/strapi/lib/core/bootstrap.js:78]: can you simply mutate them in a forEach loop ? this would be more readable I think

**Comment by derrickmehaffy** [comment]
@Convly @alexandrebodin plugin models can still be overwritten to allow changing from the default connection correct?

File: packages/strapi/lib/core/bootstrap.js (hunk 5 above)
Comment (Convly) [reply packages/strapi/lib/core/bootstrap.js]: 5d8aab2dbf7059235fac5f563a2e50e0f4734442

File: packages/strapi/lib/Strapi.js (hunk 4 above)
Comment (Convly) [reply packages/strapi/lib/Strapi.js]: 8a228b3195505acba0ab231c206506c8844b515c

File: packages/strapi-generate-model/templates/bookshelf/model.settings.template (hunk 2 above)
Comment (Convly) [reply packages/strapi-generate-model/templates/bookshelf/model.settings.template:3]: 632432bbc5342ae7c54834fa68c37a23aa8412e6

File: packages/strapi-generate-model/templates/mongoose/model.settings.template (hunk 3 above)
Comment (Convly) [reply packages/strapi-generate-model/templates/mongoose/model.settings.template:3]: 632432bbc5342ae7c54834fa68c37a23aa8412e6

File: packages/strapi-generate-model/lib/before.js (hunk 1 above)
Comment (Convly) [reply packages/strapi-generate-model/lib/before.js:125]: 632432bbc5342ae7c54834fa68c37a23aa8412e6

File: packages/strapi/lib/core/bootstrap.js [hunk 6]
```diff
@@ -75,17 +75,10 @@ module.exports = function(strapi) {
...
-    return {
-      ...acc,
-      [componentName]: _.merge(
-        // Ensure each component has a valid connection
-        { connection: component.connection || defaultConnection },
-        component
-      ),
-    };
-  }, {});
+    strapi.components[componentName].connection = component.connection || defaultConnection;
```
Comment (alexandrebodin) [inline packages/strapi/lib/core/bootstrap.js:80]: you can do component.connection = ... ;)

Comment (Convly) [reply packages/strapi/lib/core/bootstrap.js]: 8a30673334c6755bf1b9fb4d44c2c937b3a16d56

**Comment by Convly** [comment]
> @Convly @alexandrebodin plugin models can still be overwritten to allow changing from the default connection correct?

It should still work, yes

**Comment by Convly** [comment]
> I think you can just remove the `tpl` options in the model generator command now that we do not do any difference.

Good catch, done. I've updated the doc too

**Pull request merged**